esmvalcore/preprocessor/_multimodel.py

```diff
@@ -248,115 +128,179 @@ def _unify_time_coordinates(cubes):
-        time_data = _get_time_slice(cubes, time)
-        stats_data[i] = _compute_statistic(time_data, statistic)
 
-    template = cubes[0]
-    stats_cube = _put_in_cube(template, stats_data, statistic, new_times)
-    return stats_cube
+def _extend(cube, time_points):
+    """Extend cube to a specified time range.
 
+    If time points are missing before the start/after the end of the
+    time range, cubes for each missing time pointwith masked data will
+    be added to pad the time range to match `time_points`. This method
+    supports lazy operation.
+    """
+    cube.coord('time').bounds = None
+    cube_points = cube.coord('time').points
 
-def _multicube_statistics(cubes, statistics, span):
-    """Compute statistics over multiple cubes.
+    begin = cube_points[0]
+    end = cube_points[-1]
 
-    Can be used e.g. for ensemble or multi-model statistics.
+    pad_begin = time_points[time_points < begin]
+    pad_end = time_points[time_points > end]
 
-    This function was designed to work on (max) four-dimensional data:
-    time, vertical axis, two horizontal axes.
+    if (len(pad_begin)) == 0 and (len(pad_end) == 0):
+        return cube
 
-    Apart from the time coordinate, cubes must have consistent shapes. There
-    are two options to combine time coordinates of different lengths, see
-    the `span` argument.
+    template_cube = cube[:1].copy()
+    template_cube.data = np.ma.array(da.zeros_like(template_cube.data),
+                                     mask=True,
+                                     dtype=template_cube.data.dtype)
 
-    Parameters
-    ----------
-    cubes: list
-        list of cubes over which the statistics will be computed;
-    statistics: list
-        statistical metrics to be computed. Available options: mean, median,
-        max, min, std, or pXX.YY (for percentile XX.YY; decimal part optional).
-    span: str
-        overlap or full; if overlap, statitsticss are computed on common time-
-        span; if full, statistics are computed on full time spans, ignoring
-        missing data.
+    cube_list = []
 
-    Returns
-    -------
-    dict
-        dictionary of statistics cubes with statistics' names as keys.
+    for time_point in pad_begin:
+        new_slice = template_cube.copy()
+        new_slice.coord('time').points = float(time_point)
+        cube_list.append(new_slice)
 
-    Raises
-    ------
-    ValueError
-        If span is neither overlap nor full.
-    """
-    if len(cubes) < 2:
-        logger.info('Found only 1 cube; no statistics computed for %r',
-                    list(cubes)[0])
-        return {statistic: cubes[0] for statistic in statistics}
+    cube_list.append(cube)
+
+    for time_point in pad_end:
+        new_slice = template_cube.copy()
+        new_slice.coord('time').points = float(time_point)
+        cube_list.append(new_slice)
+
+    cube_list = iris.cube.CubeList(cube_list)
+
+    new_cube = cube_list.concatenate_cube()
```

Comment:
and here as well, a try/except would be better to have, exception raising a nicely explained concatenation error (iris.ConcatenationError)

Reply:
Done 👍